sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala

```diff
@@ -116,6 +119,39 @@ trait CodegenSupport extends SparkPlan {
   }
 
   /**
+    * Returns source code to evaluate all the variables, and clear the code of them, to prevent
```

Comment:
Can you add a high level comment that describes the overall framework? I think the important things to include are:
 - how it works in general?
 - how should an operator that does not short circuit (e.g. project/sort) use this?
 - how should an operator that does short circuit use this (if different)?